### PR TITLE
chore(components): improve type for popper effect

### PR DESCRIPTION
**Pull request by wzc520pyfm**
Abort #17758 #17751 #17040 #17030

At #17751 , we are forced to set the props type related to `popperEffect` to `PopperEffect | string`, causing `popperEffect ` to be deduced as `string`.

But now, here is a better solution to improve the ts type of `popperEffect`! It preserves the type hints for `light` and `drak` and accepts any string. The only downside is that it cannot pass the eslint check.

image

image

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by wzc520pyfm** [comment]
`{}` may be replaced by  `NonNullable<unknown>`, i will try it, tonight

File: packages/components/menu/src/menu.ts [hunk 1]
```diff
@@ -139,7 +139,9 @@ export const menuProps = buildProps({
    * @description Tooltip theme, built-in theme: `dark` / `light` when menu is collapsed
    */
   popperEffect: {
-    type: definePropType<PopperEffect | string>(String),
+    type: definePropType<PopperEffect | (string & NonNullable<unknown>)>(
+      String
+    ),
```
Comment (tolking) [inline packages/components/menu/src/menu.ts:144]:
```suggestion
    type: definePropType<PopperEffect>(String),
```

```ts
export type PopperEffect = typeof effects[number] | (string & {})
```

**Review by tolking — COMMENTED** [review]
I found that `(string & NonNullable<unknown>)` and `(string & {})` are the same, maybe we can use `(string & {})` directly

**Comment by wzc520pyfm** [comment]
> I found that `(string & NonNullable<unknown>)` and `(string & {})` are the same, maybe we can use `(string & {})` directly

`{}` cannot pass the eslint check.

**Comment by tolking** [comment]
> > I found that `(string & NonNullable<unknown>)` and `(string & {})` are the same, maybe we can use `(string & {})` directly
> 
> `{}` cannot pass the eslint check.

I see

**Comment by btea** [comment]
Is it possible to extract a common type from so many identical types?

File: packages/components/menu/src/menu.ts (hunk 1 above)
Comment (wzc520pyfm) [reply packages/components/menu/src/menu.ts]: Agree, I have updated.

**Comment by wzc520pyfm** [comment]
> Is it possible to extract a common type from so many identical types?

Agree, I have updated.

**Review by btea — APPROVED** [review]

**Review by tolking — APPROVED** [review]

**Pull request merged**